doc/grpc-polling-engines.md

```diff
@@ -0,0 +1,154 @@
+  - Poll the fds in the pollset for events AND return when ONE of the following is true:
+    - Deadline expired
+    - Some fds in the pollset were found to be readable/writable/error and those associated closures were ‘scheduled’ (but not necessarily executed)
+    - worker is “kicked” (see grpc_pollset_kick for more details)
+
+- **grpc_pollset_kick**
+ - Signature: `grpc_pollset_kick(grpc_pollset* ps, grpc_pollset_worker* worker)`
+ - “Kick the worker” i.e Force the worker to return from grpc_pollset_work()
+ - If `worker == nullptr`, kick ANY worker active on that pollset
+
+#### grpc_pollset_set
+
+- **grpc\_pollset\_set\_[add|del]\_fd**
+  - Signature: `grpc_pollset_set_[add|del]_fd(grpc_pollset_set* pss, grpc_fd *fd)`
+Add/Remove fd to the pollset_set
+
+- **grpc\_pollset\_set_[add|del]\_pollset**
+  - Signature: `grpc_pollset_set_[add|del]_pollset(grpc_pollset_set* pss, grpc_pollset* ps)`
+  - What does adding a pollset to a pollset_set mean ?
+    - It means that calling `grpc_pollset_work()` on the pollset will also poll all the fds in the pollset_set i.e semantically, it is similar to adding all the fds inside pollset_set to the pollset. 
+    - This guarantee is no longer true once the pollset is removed from the pollset_set
+
+- **grpc\_pollset\_set_[add|del]\_pollset\_set**
+  - Signature: `grpc_pollset_set_[add|del]_pollset_set(grpc_pollset_set* bag, grpc_pollset_set* item)`
+  - Semantically, this is similar to adding all the fds in the ‘bag’ pollset_set to the ‘item’ pollset_set
+
+
+#### Recap:
+
+__Relation between grpc_pollset_worker, grpc_pollset and grpc_fd:__
+
+![image](images/grpc-ps-pss-fd.png)
+
+__grpc_pollset_set__
+
+![image](images/grpc-pss.png)
+
+
+## Polling Engine Implementations
+
+### epoll1
+
+![image](images/grpc-epoll1.png)
+
+Code at `src/core/lib/iomgr/ev_epoll1_posix.cc`
+
+- The logic to choose a designated poller is quite complicated. Pollsets are internally sharded into what are called `pollset_neighborhood` (a structure internal to `epoll1` polling engine implementation). `grpc_pollset_workers` that call `grpc_pollset_work` on a given pollset are all queued in a linked-list against the `grpc_pollset`.
+
+- There are as many neighborhoods as the number of cores. A pollset is put in a neighborhood based on the CPU core, the root worker thread (i.e the head of the linked-list of workers queued against the pollset). The whole idea here is that when choosing the next designated poller, we make a best-effort attempt to pick a worker that is NOT running on the same core. This way, we reduce the probability of the current thread being pre-empted by the CPU scheduler.
```

Comment:
I don't understand the bit about the root worker thread since it's a sentence fragment; can you add more detail?

Also , what's the "current thread" that we're trying not to get pre-empted? I thought that all threads in this process are application threads anyway, so wouldn't the kicked thread be lost to the iomgr anyway?

Reply:
I removed the detail about thread preemption. Its a bit confusing and its a micro-optimization whose benefits are not really evident :)

Here is what I meant by 'current thread being preempted by cpu scheduler'

 - Assume there are three threads T1, T2 and T3.  T1 and T2 are scheduled on `core1` and T3 on `core2`
- Lets say T1 is designated poller. Which means T2 and T3 are 'sleeping' (i.e waiting on `gpr_cv_wait()`)
- If T1 made T2 the designated poller (i.e.., wokeup T2 from `gpr_cv_wait()`) it would mean that that now T1 an T2 are in ready-to-run- state on `core1` (and hence higher probability of T1 getting pre-empted by the cpu to give T2 a chance to run).